Serdes doc refactoring

smccarthy-ie opened the pull request:
Based on review feedback from Fabian and Eric and from looking at the updated v2 code base and example applications

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-avro.adoc:
```diff
@@ -4,43 +4,41 @@
-* Artifact ID strategy 
+* {registry} URL
+* Artifact resolver strategy 
 * Global ID strategy 
```
@smccarthy-ie sorry I think I did not mention this in the previous document, but there is no Global ID strategy anymore. That concept does not exist now, it's been replaced in part with the SchemaResolver. And the features that all our different GlobalIdStrategies had are now replaced with config properties of DefaultSchemaResolver

smccarthy-ie replied:
@famartinrh Thanks, I removed the main section on global ID strategy and replaced with default schema resolver, but wasn't sure what to do with minor references like this. Removing them all now...

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-avro.adoc:
```diff
@@ -4,43 +4,41 @@
+* Artifact resolver strategy 
 * Global ID strategy 
 * Global ID location
 * Global ID handler
```
There is no global id handler nor global id location concept. At least not with "Global Id" name :) . Depending on the location of the artifact id info ( headers or message payload) you can configure an Id Handler or a headers handler.

smccarthy-ie replied:
OK, renaming these options to more generic "ID location" and "ID encoding".

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-avro.adoc:
```diff
@@ -49,42 +47,45 @@ Avro provides different datum writers and readers to write and read data. {regis
-the message itself. In the case of the two strategies, they are not needed because the serializer is responsible for sending the global ID of the schema as part of the message. 
+The deserializer class can determine the values for these options from the message. The strategies are not required because the serializer is responsible for sending the global ID as part of the message. 
+
+The location of the global ID is determined by checking for the magic byte at the start of the message payload. If that byte is found, the global ID is read from the message payload using the configured handler.  If the magic byte is not found, the global ID is read from the message headers.
```
very good :)

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-constants.adoc:
```diff
@@ -2,73 +2,82 @@
-   public static final String REGISTRY_CONFLUENT_ID_HANDLER_CONFIG_PARAM = "apicurio.registry.as-confluent"; <4>
+   public static final String REGISTRY_URL = "apicurio.registry.url"; <1>
+   public static final String ID_HANDLER = "apicurio.registry.id-handler"; <2>
+   public static final String REGISTRY_CONFLUENT_ID_HANDLER = "apicurio.registry.as-confluent"; <3>
```
rename `REGISTRY_CONFLUENT_ID_HANDLER` to `ENABLE_CONFLUENT_ID_HANDLER`

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-constants.adoc:
```diff
@@ -2,73 +2,82 @@
+public class SerdeConfig {
+    
+   public static final String ARTIFACT_RESOLVER_STRATEGY = "apicurio.registry.artifact-resolver-strategy"; <1>
+   public static final String REGISTRY_GLOBAL_ID_STRATEGY = "apicurio.registry.global-id"; <2>
```
there is no global id strategy, this line and the referenced ones could be removed.
As I think I mentioned before, now we have `SchemaResolver` instead of global id strategy, but we want to encourage users to don't specify a Schema Resolver so the `DefaultSchemaResolver` will be used and then users will use the config properties that `DefaultSchemaResolver` accepts (see `io.apicurio.registry.serde.config.DefaultSchemaResolverConfig`) in order to do the things they did with the old `GlobalIdStrategy` and more

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-constants.adoc:
```diff
@@ -2,73 +2,82 @@
-<1> xref:service-registry-concepts-artifactid-{context}[ArtifactId strategy].
-<2> xref:service-registry-concepts-globalid-{context}[Global ID strategy].
+<1> See xref:service-registry-concepts-artifactid-registry[].
+<2> See xref:configuring-globalid-strategy-registry[].
```
as mentioned before, remove or reword this

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-json.adoc:
```diff
@@ -4,39 +4,41 @@
-* Global ID strategy
+* {registry} URL
+* Artifact resolver strategy 
+* Global ID strategy 
```
same as before, there is no global id strategy

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-json.adoc:
```diff
@@ -4,39 +4,41 @@
 The only non-standard configuration property is whether JSON Schema validation is enabled or
-disabled.  The validation feature is disabled by default but can be enabled by setting
-`apicurio.registry.serdes.json-schema.validation-enabled` to `"true"`. For example:
+disabled.  The validation feature is disabled by default but you can enable this by setting
```
now it is enabled by default , check `SerdeConfig.VALIDATION_ENABLED_DEFAULT`

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-json.adoc:
```diff
@@ -4,39 +4,41 @@
 You can configure the JSON Schema deserializer class in the following ways:
 
-* {registry} location as a URL
+* {registry} URL
```
for the json deserializer there are new configuration options, check `io.apicurio.registry.serde.jsonschema.JsonSchemaKafkaDeserializerConfig`

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-protobuf.adoc:
```diff
@@ -5,43 +5,52 @@
-* Global ID strategy 
+* {registry} URL
+* Artifact resolver strategy 
+* Global ID strategy  
```
same as before, no longer exists

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-protobuf.adoc:
```diff
@@ -5,43 +5,52 @@
+* Global ID strategy  
 * Global ID location
 * Global ID handler
 
```
there are new config options for protobuf serializer, check `io.apicurio.registry.serde.protobuf.ProtobufKafkaSerializerConfig`.
In particular, the new option is "apicurio.registry.serde.validation-enabled" (same as for json schema serde)

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-protobuf.adoc:
```diff
@@ -5,43 +5,52 @@
+You must configure the Protobuf deserializer class to match the following configuration settings in the serializer:
 
 * {registry} location as a URL
 * Global ID handler
```
global Id handler is not a configuration option for protobuf deserializer.
Besides that there are some new config properties , check `io.apicurio.registry.serde.protobuf.ProtobufKafkaDeserializerConfig`

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-protobuf.adoc:
```diff
@@ -5,43 +5,52 @@
-* Artifact ID strategy
-* Global ID strategy
+* Artifact resolver strategy 
+* Global ID strategy 
```
same as before, remove, no longer exists

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-strategy.adoc:
```diff
@@ -48,31 +48,48 @@ However, you can use alternative conventions for the mapping by using a strategy
 == Strategies to return a global ID
 
-{registry} provides the following strategies to return a global ID based on an implementation of `GlobalIdStrategy`:
+When using the `DefaultSchemaResolver`, you can configure the following application properties for the global ID strategy: 
```
I would change "for the global ID strategy" with something like "to adjust the schema resolver behavior to your needs" .

famartinrh commented on docs/modules/ROOT/partials/getting-started/con-registry-serdes-strategy.adoc:
```diff
@@ -48,31 +48,48 @@ However, you can use alternative conventions for the mapping by using a strategy
-[id='configuring-globalid-strategy-{context}']
-== Global ID strategy configuration
-You can configure the following application property for the global ID strategy:
+|===
```
there are more properties, check `io.apicurio.registry.serde.config.DefaultSchemaResolverConfig`.

famartinrh commented:
very good overall, I added a bunch of comments with things I would change.

Just one more comment.
We may need a section where we document the new fallback artifact functionality , it is a deserializer specific configuration (and applies to all data formats, it's data format agnostic actually) . There is a new interface `FallbackArtifactProvider` but we provide a  `DefaultFallbackArtifactProvider` with simple config properties. Check `io.apicurio.registry.serde.config.BaseKafkaDeserializerConfig` and `SerdeConfig.FALLBACK_ARTIFACT_PROVIDER` and the mentioned classes (there is some javadoc that may help)

smccarthy-ie commented:
Thanks @famartinrh The latest commit addresses your feedback. 

BTW, I didn't include a new section on the new "fallback" config options, but have included several links to the SerdeConfig class, which describes all the available config options.

EricWittmann commented:
@smccarthy-ie It looks like you've incorporated or at least responded to all the comments.  Is this ready to merge now?

smccarthy-ie commented:
@EricWittmann yes, ready to go. Can create a fresh PR if there are any more changes to Serdes. Thanks.
